### PR TITLE
Plans: Add free trials without going through checkout

### PR DESCRIPTION
**Pull request by drewblaisdell**
This PR updates the components on `/plans` such that a free trial is added to the user's account when they click 'Start Free Trial' without going through checkout.

**Testing**
- [x] *Free trials on `/plans`*
- Run `localStorage.setItem( 'ABTests', '{ "freeTrials_20160120":"offered" }' );`
- Visit `/plans/:site` for a site that has not had a free trial before.
- Click 'Start Free Trial' under one of the plans.
- Assert that the buttons to select a plan are disabled.
- Assert that a 'Submitting' notice appears.
- Assert that you are redirected to `/plans/:site/thank-you`.

- [x] *Plans on `/plans` with free trials enabled*
- Run `localStorage.setItem( 'ABTests', '{ "freeTrials_20160120":"offered" }' );`
- Visit `/plans/:site` for a site that has not had a free trial before.
- Assert that you can add a full plan to your cart by clicking the 'upgrade now' link under the 'Start Free Trial' buttons.
- Assert that you can complete checkout with this plan successfully.

- [x] *Plans on `/plans` with free trials disabled*
- Run `localStorage.setItem( 'ABTests', '{ "freeTrials_20160120":"notOffered" }' );`
- Visit `/plans/:site`.
- Assert that you can complete checkout after adding a plan to your cart by clicking 'Upgrade Now'.

- [x] *Free plan in signup with free trials enabled*
- Run `localStorage.setItem( 'ABTests', '{ "freeTrials_20160120":"offered" }' );`
- Visit `/start`
- Proceed through signup, selecting the free plan.
- Assert that you are redirected to `[site].wordpress.com` after the signup is complete.

- [x] *Paid plans in signup with free trials enabled*
- Run `localStorage.setItem( 'ABTests', '{ "freeTrials_20160120":"offered" }' );`
- Visit `/start`
- Proceed through signup, selecting a paid plan.
- Assert that you are redirected to checkout after signup with the plan you selected in your cart.

- [x] *Free plan in signup with free trials disabled*
- Run `localStorage.setItem( 'ABTests', '{ "freeTrials_20160120":"notOffered" }' );`
- Visit `/start`
- Proceed through signup, selecting the free plan.
- Assert that you are redirected to `[site].wordpress.com` after the signup is complete.

- [x] *Paid plans in signup with free trials disabled*
- Run `localStorage.setItem( 'ABTests', '{ "freeTrials_20160120":"notOffered" }' );`
- Visit `/start`
- Proceed through signup, selecting a paid plan.
- Assert that you are redirected to checkout after signup with the plan you selected in your cart.

- [x] Product review
- [x] Code review

### CODE REVIEW

File: client/lib/upgrades/actions/checkout.js [hunk 1]
```diff
@@ -29,6 +35,21 @@ function setNewCreditCardDetails( options ) {
...
+	const addFunction = cartItems.add( cartItem );
+
+	let cart = cartValues.emptyCart( site.ID );
+
+	cart = cartValues.fillInAllCartItemAttributes( addFunction( cart ), productsList.get() );
+
+	wpcom.undocumented().transactions( 'POST', {
+		cart,
+		payment: {
+			paymentMethod: 'WPCOM_Billing_WPCOM'
```
Comment (Tug) [inline client/lib/upgrades/actions/checkout.js]: You can replace this with `storeTransactions.fullCreditsPayment()` ;)

Comment (drewblaisdell) [inline client/lib/upgrades/actions/checkout.js]: Thanks!

File: client/components/plans/plan-actions/index.jsx [hunk 2]
```diff
@@ -145,6 +151,18 @@ module.exports = React.createClass( {
 
 		if ( this.props.onSelectPlan ) {
 			this.props.onSelectPlan( cartItem );
```
Comment (scruffian) [inline client/components/plans/plan-actions/index.jsx]: Are there cases where this will still be used?

File: client/components/plans/plan-actions/index.jsx [hunk 3]
```diff
@@ -119,6 +123,8 @@ module.exports = React.createClass( {
 	},
 
 	handleAddToCart: function( cartItem, actionType, event ) {
```
Comment (scruffian) [inline client/components/plans/plan-actions/index.jsx]: Should we change this function name?

File: client/components/plans/plan-actions/index.jsx [hunk 4]
```diff
@@ -145,6 +151,18 @@ module.exports = React.createClass( {
 
 		if ( this.props.onSelectPlan ) {
 			this.props.onSelectPlan( cartItem );
+		} else if ( cartItem.free_trial ) {
+			upgradesActions.submitFreeTransaction( this.props.site, cartItem, function( error ) {
+				if ( ! error ) {
```
Comment (scruffian) [inline client/components/plans/plan-actions/index.jsx]: What about if there is an error?

File: client/lib/upgrades/actions/checkout.js [hunk 5]
```diff
@@ -29,6 +35,21 @@ function setNewCreditCardDetails( options ) {
 	} );
 }
 
+function submitFreeTransaction( site, cartItem, onComplete ) {
```
Comment (Tug) [inline client/lib/upgrades/actions/checkout.js]: I think it would be moe consistent to give a `cart` instead of a `cartItem` here

Comment (Tug) [inline client/lib/upgrades/actions/checkout.js]: Also I'm not sure `lib/upgrades/actions/` is the best place to put our new code. It seems it defines Flux actions only.

File: client/components/plans/plan-actions/index.jsx (hunk 2 above)
Comment (drewblaisdell) [inline client/components/plans/plan-actions/index.jsx]: Yes, in signup.

File: client/lib/upgrades/actions/checkout.js (hunk 5 above)
Comment (drewblaisdell) [inline client/lib/upgrades/actions/checkout.js]: I updated this so that it is associated with a Flux action by way of `submitTransaction`. We can probably get rid of it entirely now, in favor of calling `upgradesActions.submitTransaction` directly in `PlanActions`.

File: client/lib/upgrades/notices.jsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+export function displayError( error ) {
+	if ( typeof error.message === 'object' ) {
+		notices.error( <ValidationErrorList messages={ flatten( values( error.message ) ) } /> );
+	} else {
+		notices.error( getErrorFromApi( error.message ) )
+	}
+}
+
+export function displaySubmitting( isFreeCart ) {
```
Comment (drewblaisdell) [inline client/lib/upgrades/notices.jsx]: @Tug this method signature is a [boolean trap](http://ariya.ofilabs.com/2011/08/hall-of-api-shame-boolean-trap.html), which makes for confusing code. We sometimes circumvent this issue by using an object (which the destructuring syntax makes even easier):

```js
function displaySubmitting( { isFreeCart } ) {
```

File: client/lib/upgrades/actions/free-checkout.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ * Internal dependencies
+ */
+import { cartItems, emptyTemporaryCart, fillInAllCartItemAttributes } from 'lib/cart-values';
+import productsListFactory from 'lib/products-list';
+import { fullCreditsPayment } from 'lib/store-transactions';
+import { submitTransaction } from './checkout';
+
+const productsList = productsListFactory();
+
+export function freeTrial( siteID, plan, onComplete ) {
```
Comment (drewblaisdell) [inline client/lib/upgrades/actions/free-checkout.js]: @Tug Good idea to move this out of `actions/checkout.js`. What do you think about naming it `actions/free-trials.js`, and naming this method `startFreeTrial`? My rationale for these changes:

- Right now, this is all free-trial-specific. Some "free checkouts" (purchases made with credits) still go through checkout.
- We generally include a verb with actions.

File: client/lib/upgrades/actions/free-checkout.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+	const emptyCart = emptyTemporaryCart( siteID ),
+		planItem = cartItems.getItemForPlan( plan, { isFreeTrial: true } ),
+		cart = cartItems.add( planItem )( emptyCart );
+
+	submitFreeTransaction( cart, onComplete )
+}
+
+export function submitFreeTransaction( partialCart, onComplete ) {
```
Comment (drewblaisdell) [inline client/lib/upgrades/actions/free-checkout.js]: @Tug I don't believe we need to export this function.

File: client/components/plans/plan-actions/index.jsx [hunk 9]
```diff
@@ -2,26 +2,42 @@
...
+	puchasesPaths = require( 'me/purchases/paths' ),
+	refreshSitePlans = require( 'state/sites/plans/actions' ).refreshSitePlans,
+	upgradesActions = require( 'lib/upgrades/actions' ),
+	upgradesNotices = require( 'lib/upgrades/notices' );
 
+var PlanActions = React.createClass( {
 	propTypes: { plan: React.PropTypes.object },
 
+	componentWillReceiveProps: function( nextProps ) {
+		if ( nextProps.isSubmitting ) {
```
Comment (drewblaisdell) [inline client/components/plans/plan-actions/index.jsx]: @Tug This component has a lot of props, so we should probably guard against situations where other prop changes would trigger multiple `displaySubmitting` calls like you did below:

```js
if ( ! this.props.isSubmitting && nextProps.isSubmitting ) {
```

Comment (drewblaisdell) [inline client/components/plans/plan-actions/index.jsx]: @Tug Also, why not just call this as soon as the user selects a plan, in `handleStartFreeTrial`?

File: client/components/plans/plan-actions/index.jsx [hunk 10]
```diff
@@ -95,57 +112,77 @@ module.exports = React.createClass( {
...
-			this.props.onSelectPlan( cartItem );
+			return this.props.onSelectPlan( cartItem );
 		}
+
+		upgradesActions.addItem( cartItem );
+
+		page( '/checkout/' + this.props.site.slug );
+	},
+
+	handleStartFreeTrial: function( event ) {
```
Comment (drewblaisdell) [inline client/components/plans/plan-actions/index.jsx]: @Tug This is super minor, but I think we should only prefix methods with `handle` that relate to the event to which they are bound, e.g. `handleClickStartTrial`. Ideally, we have methods that are simple enough that we can just set event handlers like `onClick={ this.startTrial }`, but this method also handles recording the click, so I think using a handle-prefixed method is appropriate.

File: client/lib/cart-values/index.js [hunk 11]
```diff
@@ -16,6 +16,10 @@ function emptyCart( siteID ) {
 	return { blog_id: siteID, products: [] };
 }
 
+function emptyTemporaryCart( siteID ) {
```
Comment (drewblaisdell) [inline client/lib/cart-values/index.js]: @Tug Before we set the temporary flag, was the user's cart cleared when starting a free trial?

What do you think about either calling `emptyCart` from this method, or extending it to use an optional `attributes` object?

```js
function emptyCart( siteId, attributes ) {
  return Object.assign( { blog_id: siteId, products: [], attributes } );
}

var cart = emptyCart( { temporary: true } );
```

**Comment by drewblaisdell** [comment]
Those commits look good, @tug. I added some notes, and will add commits to resolve them.

File: client/components/plans/plan-actions/index.jsx [hunk 12]
```diff
@@ -95,57 +112,77 @@ module.exports = React.createClass( {
...
-			}
+		const actionType = event.target.tagName === 'A' ? 'link' : 'button';
+
+		if ( 'link' === actionType ) {
+			this.recordUpgradeNowClick();
+		} else {
+			this.recordUpgradeNowButton();
 		}
 
+		const cartItem = cartItems.getItemForPlan( this.props.plan );
```
Comment (drewblaisdell) [inline client/components/plans/plan-actions/index.jsx]: @Tug There isn't a cart item for `free_plan`, so we need to check for that before calling this function.

**Comment by drewblaisdell** [comment]
@Tug I added a commit/updated your commits to address my notes. Let me know if you're cool with the changes, which are documented by my notes.

File: client/lib/upgrades/notices.jsx (hunk 6 above)
Comment (Tug) [inline client/lib/upgrades/notices.jsx]: Agreed, it's better :)

File: client/lib/upgrades/actions/free-checkout.js (hunk 7 above)
Comment (Tug) [inline client/lib/upgrades/actions/free-checkout.js]: Yeah it probably makes more sense to name this file `free-trials.js` while it only contains a method for free trials. I wanted to be too generic. `startFreeTrial` is better also

File: client/lib/upgrades/actions/free-checkout.js (hunk 8 above)
Comment (Tug) [inline client/lib/upgrades/actions/free-checkout.js]: If we don't need it then let's hide it ;)

File: client/components/plans/plan-actions/index.jsx (hunk 9 above)
Comment (Tug) [inline client/components/plans/plan-actions/index.jsx]: Yep ok let's do this, if you find it clearer!
Oh wait, you did all the fixes already, great!

File: client/lib/cart-values/index.js (hunk 11 above)
Comment (Tug) [inline client/lib/cart-values/index.js]: Nice I like that

**Comment by Tug** [comment]
Great fixes! Thanks @drewblaisdell, I think we can start reviewing all this :)

File: client/components/plans/plan-actions/index.jsx [hunk 13]
```diff
@@ -95,57 +101,87 @@ module.exports = React.createClass( {
...
 
-			if ( actionType === 'link' ) {
-				if ( cartItem.free_trial ) {
-					this.recordUpgradeTrialNowClick( cartItem );
-				}
-				if ( ! cartItem.free_trial ) {
-					this.recordUpgradeNowClick( cartItem );
-				}
-			}
+		const actionType = event.target.tagName === 'A' ? 'link' : 'button';
```
Comment (scruffian) [inline client/components/plans/plan-actions/index.jsx]: This looks fragile.

**Comment by scruffian** [comment]
When I try to start a trial from /plans, I see this:
screen shot 2016-03-09 at 10 19 09

It would be nicer if it was all on one line...

**Comment by scruffian** [comment]
I also tested what happens when the API returns an error:
screen shot 2016-03-09 at 10 37 39

Looks good.

**Comment by scruffian** [comment]
I also tested what happens if you already have something in your cart. Steps:
1. Add a domain registration to your cart
2. Go to /plans and assert that there is a domain your cart
2. Sign up for a free trial
3. Check for the domain registration in your cart. It should still be there but in my case it was removed.

**Comment by Tug** [comment]
@scruffian I checked this already in a previous PR and this a a bug that needs to be fixed on WPCOM. At the moment, if you submit a temporary cart to the transaction endpoint it will erase the current cart (ignoring the temporary attribute)

**Comment by scruffian** [comment]
Another thing I noticed: When a user has a Premium plan, they are not given the option of a business free trial:
screen shot 2016-03-09 at 10 49 05

This is probably fine, but I wanted to flag it.

**Comment by scruffian** [comment]
I think I found another bug, with users who purchase domains. Steps to reproduce:
1. Run localStorage.setItem( 'ABTests', '{ "freeTrials_20160120":"offered" }' );
2. Visit `/plans/:site`
3. Purchase a domain
4. Return to `/plans/:site`
5. Notice that you are given the option to start a free trial
6. Refresh the page and assert that you are no longer given this option.

I am not sure whether users who have already purchased a domain should be offered a free trial here. I think they should be, but if they aren't then we should refetch the `/site/:site/plans` endpoint when the user purchases a domain, so that we don't show stale data on that page.

File: client/components/plans/plan-actions/index.jsx [hunk 14]
```diff
@@ -189,14 +229,14 @@ module.exports = React.createClass( {
 			);
 		}
 
-		if ( isFreePlan( this.props.plan ) ) {
+		if ( this.shouldOfferFreeTrial() ) {
 			return (
-				<div onClick={ this.handleAddToCart.bind( null, null, 'button' ) } className={ classes } />
+				<div onClick={ this.handleClickStartTrial } className={ classes } />
```
Comment (scruffian) [inline client/components/plans/plan-actions/index.jsx]: Is this because we have to avoid using `bind`? Another option would be to create two methods that both call this, like: `handleLinkClickStartTrial` and `handleButtonClickStartTrial`.

File: client/components/plans/plans-compare/index.jsx [hunk 15]
```diff
@@ -95,7 +95,8 @@ var PlansCompare = React.createClass( {
 					site={ site }
 					cart={ this.props.cart }
 					features={ featuresList }
-					sitePlans={ this.props.sitePlans } />
+					sitePlans={ this.props.sitePlans }
+					isSubmitting={ this.props.transaction.step.name === 'submitting-wpcom-request' } />
```
Comment (scruffian) [inline client/components/plans/plans-compare/index.jsx]: Should we make `submitting-wpcom-request` a `CONSTANT`?

File: client/lib/upgrades/notices.jsx [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+import values from 'lodash/values';
+
+/**
+ * Internal dependencies
+ */
+import i18n from 'lib/mixins/i18n';
+import notices from 'notices'
+import ValidationErrorList from 'notices/validation-error-list';
+
+function getErrorFromApi( errorMessage ) {
```
Comment (scruffian) [inline client/lib/upgrades/notices.jsx]: Thanks for moving this. I'm glad my name won't be associated with this anymore ;)

File: client/my-sites/upgrades/checkout/transaction-steps-mixin.jsx [hunk 17]
```diff
@@ -50,49 +47,21 @@ var TransactionStepsMixin = {
...
 			case 'input-validation':
 				if ( ! cartItems.hasFreeTrial( cart ) ) {
-					notices.info( isFree( cart ) ? this.translate( 'Submitting' ) : this.translate( 'Submitting payment' ) );
+					upgradesNotices.displaySubmitting( { isFreeCart: isFree( cart ) } );
 				}
 				break;
 
 			case 'received-wpcom-response':
-				notices.clearNotices( 'notices' );
+				upgradesNotices.clear();
```
Comment (scruffian) [inline client/my-sites/upgrades/checkout/transaction-steps-mixin.jsx]: Is this abstraction necessary or worth it?

**Comment by fabianapsimoes** [comment]
> When I try to start a trial from /plans, I see this:

There's a similar issue reported here: https://github.com/Automattic/wp-calypso/issues/3593

**Comment by fabianapsimoes** [comment]
> Another thing I noticed: When a user has a Premium plan, they are not given the option of a business free trial.

I suppose that's fine. It'd introduce a lot of complexity to deal with putting a subscription on hold while a user takes a trial of a different plan.

**Comment by Tug** [comment]
> Another thing I noticed: When a user has a Premium plan, they are not given the option of a business free trial:

It seems `can_start_trial` is false for the business plan in this case. So I guess this is a deliberate choice in the API.

**Comment by fabianapsimoes** [comment]
> I am not sure whether users who have already purchased a domain should be offered a free trial here. I think they should be, but if they aren't then we should refetch the /site/:site/plans endpoint when the user purchases a domain, so that we don't show stale data on that page.

Yup, I do think they should be offered a free trial in this case. We need to fix that.

**Comment by Tug** [comment]
It seems to be an API issue. I have reproduced Ben's issue with success and I have noticed I have `can_start_trial` to false for the premium plan.

File: client/components/plans/plan-actions/index.jsx (hunk 14 above)
Comment (Tug) [inline client/components/plans/plan-actions/index.jsx]: Yep we could do that. I do think that avoiding bind here makes the code easier to understand / cleaner.

File: client/components/plans/plans-compare/index.jsx (hunk 15 above)
Comment (Tug) [inline client/components/plans/plans-compare/index.jsx]: Yes, or have an helper function in one of the transaction lib such as `isSubmitting( transaction )`

File: client/my-sites/upgrades/checkout/transaction-steps-mixin.jsx (hunk 17 above)
Comment (Tug) [inline client/my-sites/upgrades/checkout/transaction-steps-mixin.jsx]: Well since `upgradesNotices` now handles show upgrades notices I guess it should also handle hiding them. This way we only import `upgradesNotices` and not `notices`

**Comment by scruffian** [comment]
> I am not sure whether users who have already purchased a domain should be offered a free trial here. I think they should be, but if they aren't then we should refetch the /site/:site/plans endpoint when the user purchases a domain, so that we don't show stale data on that page.

This depends on whether we plan on offering free trials on the /plans page. We need to get a clear answer on this question before we do the work to allow users with a domain to purchase a plan. Let's talk about it in the hangout tomorrow @breezyskies.

File: client/components/plans/plan-actions/index.jsx (hunk 14 above)
Comment (Tug) [inline client/components/plans/plan-actions/index.jsx]: I added back bind in the end...

Comment (drewblaisdell) [inline client/components/plans/plan-actions/index.jsx]: We should avoid using bind in render() because it creates a new function (slowly), making it necessary to update the DOM with every branch in the render tree containing a bind. I think that checking the event to avoid this was a reasonable fix, but this file already used bind like this and I think we can update it elsewhere.

File: client/components/plans/plans-compare/index.jsx [hunk 18]
```diff
@@ -24,7 +24,8 @@ var observe = require( 'lib/mixins/data-observe' ),
 	Card = require( 'components/card' ),
 	featuresListUtils = require( 'lib/features-list/utils' ),
 	filterPlansBySiteAndProps = require( 'lib/plans' ).filterPlansBySiteAndProps,
-	shouldFetchSitePlans = require( 'lib/plans' ).shouldFetchSitePlans;
+	shouldFetchSitePlans = require( 'lib/plans' ).shouldFetchSitePlans,
+	TransactionStepTypes = require( 'lib/store-transactions/step-types' );
```
Comment (drewblaisdell) [inline client/components/plans/plans-compare/index.jsx]: Good idea to use constants here. Also, super minor, but I think the first word should be lower case in this variable: `transactionStepTypes`

Comment (Tug) [inline client/components/plans/plans-compare/index.jsx]: Hmm, probably a bad habit from Java where I see this as an enum class ;)

**Comment by Tug** [comment]
Here you go, ready for review again :)
I feel like we should stop modifying things as this one has become quite large.
For code reviewers, please review each commit one after the other.

File: client/lib/upgrades/notices.jsx (hunk 16 above)
Comment (drewblaisdell) [inline client/lib/upgrades/notices.jsx]: We should add a `// FIXME: @scruffian` here.

**Comment by drewblaisdell** [comment]
I removed https://github.com/Automattic/wp-calypso/commit/87df2214908786a264fac7000174b5be3a20e845 per our discussion on that commit.

File: client/components/plans/plan-actions/index.jsx (hunk 13 above)
Comment (Tug) [inline client/components/plans/plan-actions/index.jsx]: I think it's fine for the moment, a click event will always have a target prop and an HTML element will always have a tagName, so it should not break. I don't think it's really fragile per say but I agree that it's not easily extensible.

File: client/lib/cart-values/index.js [hunk 19]
```diff
@@ -13,8 +13,8 @@ var cartItems = require( './cart-items' ),
 	productsValues = require( 'lib/products-values' ),
 	{ abtest } = require( 'lib/abtest' );
 
-function emptyCart( siteID ) {
-	return { blog_id: siteID, products: [] };
+function emptyCart( siteId, attributes ) {
```
Comment (stephanethomas) [inline client/lib/cart-values/index.js]: Maybe it's time to add a docblock comment for this function?

File: client/components/notice/style.scss [hunk 20]
```diff
@@ -44,7 +44,6 @@
 
 .notice__text {
 	flex-grow: 1;
-	flex-basis: 100px;
```
Comment (scruffian) [inline client/components/notice/style.scss]: I thought we were going to move this to a different PR?

Comment (Tug) [inline client/components/notice/style.scss]: I thought as well. I have just removed the corresponding commit :)

File: client/lib/cart-values/index.js (hunk 19 above)
Comment (Tug) [inline client/lib/cart-values/index.js]: Yes, done!

**Comment by scruffian** [comment]
Let's handle the side issues here in other pull requests. Code :+1:

File: client/lib/upgrades/actions/free-trials.js [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+ * Internal dependencies
+ */
+import { cartItems, emptyCart, fillInAllCartItemAttributes } from 'lib/cart-values';
+import productsListFactory from 'lib/products-list';
+import { fullCreditsPayment } from 'lib/store-transactions';
+import { submitTransaction } from './checkout';
+
+const productsList = productsListFactory();
+
+function submitFreeTransaction( partialCart, onComplete ) {
```
Comment (stephanethomas) [inline client/lib/upgrades/actions/free-trials.js]: Could we add docblock comments to these two functions please? That would really help to get a better understanding of how they operate. For example it's not clear what a partial cart is or whether a plan is an object or a string.

Comment (Tug) [inline client/lib/upgrades/actions/free-trials.js]: I agree this is needed. On it!

**Comment by scruffian** [comment]
Product :+1:
